lib/membrane/rtp/twcc_receiver/packet_info_store.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Membrane.RTP.TWCCReceiver.PacketInfoStore do
+  @moduledoc false
+
+  # The module stores TWCC sequence number along with their arrival timestamps, handling sequence
+  # number rollovers if necessary. Stored packet info can used for generating statistics used for
+  # assembling a TWCC feedback packet.
+
+  alias Membrane.Time
+
+  require Bitwise
+
+  defstruct base_seq_num: nil,
+            max_seq_num: nil,
+            seq_to_timestamp: %{}
+
+  @type t :: %__MODULE__{
+          base_seq_num: non_neg_integer(),
+          max_seq_num: non_neg_integer(),
+          seq_to_timestamp: %{non_neg_integer() => Time.t()}
+        }
+
+  @type stats_t :: %{
+          base_seq_num: non_neg_integer(),
+          packet_status_count: non_neg_integer(),
+          receive_deltas: [Time.t() | :not_received],
+          reference_time: Time.t()
+        }
+
+  @seq_number_limit 65_536
+
+  @spec empty?(__MODULE__.t()) :: boolean
+  def empty?(%__MODULE__{base_seq_num: base_seq_num}), do: base_seq_num == nil
+
+  @spec insert_packet_info(__MODULE__.t(), non_neg_integer()) :: __MODULE__.t()
+  def insert_packet_info(store, seq_num) do
+    arrival_ts = Time.vm_time()
+    {store, seq_num} = maybe_handle_rollover(store, seq_num)
+
+    %{
+      store
+      | base_seq_num: min(store.base_seq_num, seq_num) || seq_num,
+        max_seq_num: max(store.max_seq_num, seq_num) || seq_num,
+        seq_to_timestamp: Map.put(store.seq_to_timestamp, seq_num, arrival_ts)
+    }
+  end
+
+  @spec get_stats(__MODULE__.t()) :: stats_t()
+  def get_stats(store) do
+    {reference_time, receive_deltas} = make_receive_deltas(store)
+    packet_status_count = store.max_seq_num - store.base_seq_num + 1
+
+    %{
+      base_seq_num: store.base_seq_num,
+      packet_status_count: packet_status_count,
+      reference_time: reference_time,
+      receive_deltas: receive_deltas
+    }
+  end
+
+  defp maybe_handle_rollover(store, new_seq_num) do
+    %{
+      base_seq_num: base_seq_num,
+      max_seq_num: max_seq_num,
+      seq_to_timestamp: seq_to_timestamp
+    } = store
+
+    case from_which_cycle(base_seq_num, new_seq_num) do
```

Comment:
Is this the same logic as in jitter buffer element? Right now this logic is spreaded in 3 places, JitterBuffer, VAD and now here. Can we extract it to one place (only if this logic is common for all of the modules)?

Reply:
Good catch, in fact this logic was in 4 modules 😆 
Extracted to `RTP.Utils.from_which_cycle/3`